sdks/go/pkg/beam/io/bigtableio/bigtable.go

```diff
@@ -0,0 +1,278 @@
+// analogue to bigtable.Mutation.Set().
+// The timestamp will be truncated to millisecond granularity.
+// A timestamp of ServerTime means to use the server timestamp.
+func (m *Mutation) Set(family, column string, ts bigtable.Timestamp, value []byte) {
+	m.Ops = append(m.Ops, Operation{Family: family, Column: column, Ts: ts, Value: value})
+}
+
+// WithGroupKey sets a custom group key to be utilised by beam.GroupByKey.
+func (m *Mutation) WithGroupKey(key string) *Mutation {
+	m.GroupKey = key
+	return m
+}
+
+// Write writes the elements of the given PCollection<bigtableio.Mutation> to bigtable.
+func Write(s beam.Scope, project, instanceID, table string, col beam.PCollection) {
+	t := col.Type().Type()
+	err := mustBeBigtableioMutation(t)
+	if err != nil {
+		panic(err)
+	}
+
+	s = s.Scope("bigtable.Write")
+
+	pre := beam.ParDo(s, addGroupKeyFn, col)
+	post := beam.GroupByKey(s, pre)
+	beam.ParDo0(s, &writeFn{Project: project, InstanceID: instanceID, TableName: table, Type: beam.EncodedType{T: t}}, post)
+}
+
+// WriteBatch writes the elements of the given PCollection<bigtableio.Mutation> 
+// to bigtable using bigtable.ApplyBulk().
+// For the underlying bigtable.ApplyBulk function to work properly 
+// the maximum number of operations per bigtableio.Mutation of the input
+// PCollection must be given (maxOpsPerMutation). This is necessary due to 
+// the maximum amount of mutations allowed per bulk operation (100,000),
+// see https://cloud.google.com/bigtable/docs/writes#batch for more.
+func WriteBatch(s beam.Scope, project, instanceID, table string, maxOpsPerMutation uint, col beam.PCollection) {
+	t := col.Type().Type()
+	err := mustBeBigtableioMutation(t)
+	if err != nil {
+		panic(err)
+	}
+
+	if maxOpsPerMutation == 0 {
+		panic("maxOpsPerMutation must not be 0")
+	}
+
+	if maxOpsPerMutation > 100000 {
+		panic("maxOpsPerMutation must not be greater than 100,000, see https://cloud.google.com/bigtable/docs/writes#batch")
+	}
+
+	s = s.Scope("bigtable.WriteBatch")
+
+	pre := beam.ParDo(s, addGroupKeyFn, col)
+	post := beam.GroupByKey(s, pre)
+	beam.ParDo0(s, &writeBatchFn{Project: project, InstanceID: instanceID, TableName: table, MaxOpsPerMutation: maxOpsPerMutation, Type: beam.EncodedType{T: t}}, post)
+}
+
+func addGroupKeyFn(mutation Mutation) (int, Mutation) {
+	if mutation.GroupKey != "" {
+		return hashStringToInt(mutation.GroupKey), mutation
+	}
+	return 1, mutation
+}
+
+func hashStringToInt(s string) int {
+	h := fnv.New32a()
+	h.Write([]byte(s))
+	return int(h.Sum32())
+}
+
+func mustBeBigtableioMutation(t reflect.Type) error {
+	if t != reflect.TypeOf(Mutation{}) {
+		return fmt.Errorf("type must be bigtableio.Mutation but is: %v", t)
+	}
+	return nil
+}
+
+type writeFn struct {
+	// Project is the project
+	Project string `json:"project"`
+	// InstanceID is the bigtable instanceID
+	InstanceID string `json:"instanceId"`
+	// Client is the bigtable.Client
+	Client *bigtable.Client `json:"client"`
```

Comment:
Unless the client and table are JSON serializable, it's probably best to leave them as unexported fields for use during execution time. Here and in writeBatchFn.

Reply:
Check ☑️